File: core/src/androidMain/kotlin/ru/mobileup/template/core/map/internal/CameraController.kt
package ru.mobileup.template.core.map.internal

import com.yandex.mapkit.Animation
import com.yandex.mapkit.geometry.Geometry
import com.yandex.mapkit.geometry.Point
import com.yandex.mapkit.geometry.Polyline
import com.yandex.mapkit.map.CameraListener
import com.yandex.mapkit.map.CameraPosition
import com.yandex.mapkit.map.Map
import com.yandex.mapkit.mapview.MapView
import ru.mobileup.template.core.location.GeoCoordinate
import ru.mobileup.template.core.map.MapCameraPosition

private const val ANIMATION_DURATION_SECONDS = 1.0f

internal class CameraController(
    private val mapView: MapView,
    private val onCameraPositionChange: (MapCameraPosition) -> Unit
) {

    private val cameraListener = CameraListener { _, position, _, _ ->
        onCameraPositionChange(position.toMapCameraPosition())
    }

    val cameraPosition: MapCameraPosition
        get() = mapView.mapWindow.map.cameraPosition.toMapCameraPosition()

    init {
        mapView.mapWindow.map.addCameraListener(cameraListener)
    }

    fun moveCamera(
        position: MapCameraPosition,
        animate: Boolean
    ) {
        mapView.mapWindow.map.moveCamera(position.toYandexCameraPosition(), animate)
    }

    fun calculateCameraPosition(coordinates: List<GeoCoordinate>): MapCameraPosition? {
        if (coordinates.isEmpty()) return null

        val geometry = Geometry.fromPolyline(Polyline(coordinates.map { it.toPoint() }))
        val currentPosition = mapView.mapWindow.map.cameraPosition
        val position = mapView.mapWindow.map.cameraPosition(
            geometry,
            null,
            currentPosition.azimuth,
            currentPosition.tilt
        )

        return MapCameraPosition(
            coordinate = position.target.toGeoCoordinate(),
            zoom = position.zoom,
            azimuth = position.azimuth,
            tilt = position.tilt
        )
    }
}

private fun Map.moveCamera(
    position: CameraPosition,
    animate: Boolean
) {
    if (animate) {
        move(position, Animation(Animation.Type.SMOOTH, ANIMATION_DURATION_SECONDS), null)
    } else {
        move(position)
    }
}

internal fun GeoCoordinate.toPoint(): Point = Point(latitude, longitude)

private fun Point.toGeoCoordinate(): GeoCoordinate = GeoCoordinate(latitude, longitude)

private fun MapCameraPosition.toYandexCameraPosition(): CameraPosition {
    return CameraPosition(coordinate.toPoint(), zoom, azimuth, tilt)
}

private fun CameraPosition.toMapCameraPosition(): MapCameraPosition {
    return MapCameraPosition(
        coordinate = target.toGeoCoordinate(),
        zoom = zoom,
        azimuth = azimuth,
        tilt = tilt
    )
}

File: iosApp/iosApp/Source/Map/Subcontrollers/CameraController.swift
import YandexMapsMobile
import Shared

final class CameraController: NSObject, YMKMapCameraListener {

    private static let animationDurationSeconds: Float = 1.0

    private let mapView: YMKMapView
    private let onCameraPositionChange: (MapCameraPosition) -> Void

    init(
        mapView: YMKMapView,
        onCameraPositionChange: @escaping (MapCameraPosition) -> Void
    ) {
        self.mapView = mapView
        self.onCameraPositionChange = onCameraPositionChange
        super.init()
        mapView.mapWindow.map.addCameraListener(with: self)
    }

    var cameraPosition: MapCameraPosition {
        mapView.mapWindow.map.cameraPosition.mapCameraPosition
    }

    func moveCamera(position: MapCameraPosition, animate: Bool) {
        let map = mapView.mapWindow.map
        if animate {
            map.move(
                with: position.ymkCameraPosition,
                animation: YMKAnimation(type: .smooth, duration: Self.animationDurationSeconds)
            )
        } else {
            map.move(with: position.ymkCameraPosition)
        }
    }

    func calculateBoundingBoxCameraPosition(coordinates: [GeoCoordinate]) -> MapCameraPosition? {
        guard !coordinates.isEmpty else { return nil }

        let map = mapView.mapWindow.map
        let geometry = YMKGeometry(polyline: YMKPolyline(points: coordinates.map { $0.ymkPoint }))
        let currentPosition = map.cameraPosition
        let position = map.cameraPosition(
            with: geometry,
            focus: nil,
            azimuth: NSNumber(value: currentPosition.azimuth),
            tilt: NSNumber(value: currentPosition.tilt)
        )
        return position.mapCameraPosition
    }

    func dispose() {
        mapView.mapWindow.map.removeCameraListener(with: self)
    }

    // MARK: - YMKMapCameraListener

    func onCameraPositionChanged(
        with map: YMKMap,
        cameraPosition: YMKCameraPosition,
        cameraUpdateReason: YMKCameraUpdateReason,
        finished: Bool
    ) {
        onCameraPositionChange(cameraPosition.mapCameraPosition)
    }
}
